### PR TITLE
add option to disable part of speech rule check

### PR DESCRIPTION
**Pull request by StefanVukovic99**
Part 4 of #422.
Adds an option to disable part of speech rule check for deinflections.

### CODE REVIEW

File: ext/js/language/deinflector.js
Comment (Scrub1492) [inline ext/js/language/deinflector.js:83]: Extra line here

Comment (Scrub1492) [reply ext/js/language/deinflector.js:83]: Bitwise operations are inherently pretty hard to understand, can you write some comments now that we're at it? Maybe not necessarily in this PR

File: ext/data/schemas/options-schema.json
Comment (toasted-nutbread) [inline ext/data/schemas/options-schema.json:1]: Setting upgrade needed.

**Review by toasted-nutbread — COMMENTED** [review]
This is potentially me getting a bit too into the weeds on UX for the early stage of a new feature, but it seems like this property is something that is derivative of the selected language that is being translated, but correct me if I'm wrong. I'm not sure how it's used currently, or what use case it's addressing.

My reason for this comment is that I would suggest avoiding exposing internal details to users unless there is a clear need. An issue this extension already has is that there are several settings that are not strictly necessary, so I'm trying to avoid the pitfall of exposing options which may contribute to technical debt / more maintenance overhead with little added benefit.

**Comment by toasted-nutbread** [comment]
I think this has the potential to infinite loop when the part-of-speech check is disabled. Here is an example:

```
Cycle detected:
  ?して (start)
  ?してる (masu stem, =>v1, て=>てる)
  ?して (progressive or perfect, v1=>iru, てる=>て)
  ?してる (masu stem, =>v1, て=>てる) -- Cycle!
```

#471

**Comment by StefanVukovic99** [comment]
On closer inspection, changes to the deinflector may not be necessary so I've undone them. The primary purpose of this is to reduce the "barrier to entry" to making and using new dictionaries, since they, especially for other languages, may have missing/incomplete/inconsistent parts of speech. This will also help with developing deinflection for other languages, since dictionaries won't have to be updated with every change to the deinflector/rules.

File: ext/js/data/options-util.js
Comment (toasted-nutbread) [inline ext/js/data/options-util.js:1157]: no return needed

**Comment by djahandarie** [comment]
So is the idea that a dictionary developer would turn this off while making a new dictionary, and then turn it back on when the dictionary has reached a point of having good part of speech information?

I kind of wonder if this should actually be an attribute of the dictionary itself which can cause the deinflector to behave differently in this way as opposed to a profile setting. What do you think?

**Comment by StefanVukovic99** [comment]
> So is the idea that ....

That could be one way of using it. What tends to happen is dictionaries get converted with pyglossary or such, resulting in a rudimentary dictionary which only has the terms and definitions. Here's one user's english dict:
image

The deinflection rules for english which have rulesOut won't work with a dictionary which has no rules. Or, for example, the entry for 'hammer' might be marked as a noun but the definitions also cover the verb senses, but the verb deinflections won't work. 

> ... an attribute of the dictionary ...
Analogous to the current tag_bank a dictionary might also declare the rules it is using, but demanding this would just increase the burden on laymen who are trying to convert dictionaries. 

Basically, this makes the deinflector-dictionary coupling more robust, and gives users more control over their experience.

**Comment by Scrub1492** [comment]
I think its better to abstract more stuff to the dictionary level than to expose it to the users, and try to be conservative about what we present to the users. As we move forward with other languages support, there might be many more options changes, and its worth cutting down what isn't really needed.

> ... gives users more control over their experience.
Yomitan does this for a lot of stuff, and its one of the most unloved parts. Users often don't like debugging custom CSS, handlebars, or a setting; they just want the product to work out of the box.

> demanding this would just increase the burden on laymen who are trying to convert dictionaries.
I don't like the idea of moving the burden of figuring out this functionality from dictionary makers and shifting it on users. 
In addition, good dictionaries will come out with time. On the other hand, its hard to roll back settings changes that already made its way to the users. I don't think its worthwhile trying to maximize dictionary compatibility here.

**Comment by StefanVukovic99** [comment]
> I think its better to abstract more stuff to the dictionary level...

Yeah, this is probably better tied to the dictionary. I tend to use only one term dictionary so having a profile setting has been ok for me, but lots of users like to have a bunch of dictionaries, so this should probably be a per-dictionary thing rather than a blanket setting. 

That said I think this differs from the handlebars/css problem, where users basically have to maintain their own code. We are already having the user answer some questions about his dictionaries and their features:

- Which frequency dictionary to use for sorting?
- Is it rank-based or frequency-based?
- What dictionaries are primary and secondary (group related terms)?
- Which dictionaries have overly long definitions that should be collapsed?

Tactics of minimizing setup hassle should be used:
1. Defaults;
2. Auto settings (like detecting rank vs freq-based) - could check what percentage of terms have no rules
3. index.json - could support the dictionary explicitly recommending its own settings;

but the user should be able to choose manually. I think a modal like Collapsible Dictionaries, allowing the user to choose which dictionaries should have filtering, may be the best approach:
image
<sub>(My next PR will also need a similar setting with regard to which dictionaries' will have redirects for deinflection enabled, and this approach seems like a good fit for that also)</sub>

Being an advanced setting, it shouldn't be in the way for users that don't need it.

**Comment by Scrub1492** [comment]
> That said I think this differs from the handlebars/css problem, where users basically have to maintain their own code. 

Take your example of the rudimentary english dictionary: If the user of this dictionary has set this setting, and the dictionary maker updates this dictionary to have parts of speech, then if the user updates to the newer version of this dictionary things will break. If the setting is embedded inside the dictionary, things will keep working normally. This is maintainance overhead for users.

> We are already having the user answer some questions about his dictionaries and their features:
> * Which frequency dictionary to use for sorting?
> * Is it rank-based or frequency-based?
> * What dictionaries are primary and secondary (group related terms)?
> * Which dictionaries have overly long definitions that should be collapsed?

These are all preference-related questions. This setting can break yomitan lookups (or severely hurt performance).

> Tactics of minimizing setup hassle should be used:
> 
> 1. Defaults;
> 2. Auto settings (like detecting rank vs freq-based) - could check what percentage of terms have no rules
> 3. index.json - could support the dictionary explicitly recommending its own settings;
> 
> but the user should be able to choose manually. I think a modal like Collapsible Dictionaries, allowing the user to choose which dictionaries should have filtering, may be the best approach: image (My next PR will also need a similar setting with regard to which dictionaries' will have redirects for deinflection enabled, and this approach seems like a good fit for that also)

Again, override settings may break things if the user were to update a dictionary.


That said, I think you have a better idea of what the finished product looks like, and I'm not trying to seem picky; just that a major update like this needs to be treaded with caution.

**Comment by djahandarie** [comment]
So I think it sounds like we have agreement that we should have more dictionary-level settings, and this setting would likely be more appropriate as such. Such settings should likely be included with the dictionary in `index.json`, and also be exposed in the UI so advanced users can modify them.

We in fact already have a dictionary settings UI, if you go to `Settings > Dictionaries > Configure > ⋮ on a dictionary of choice > Details`. I would suggest that we start by putting things in there.

That said, if we start putting important stuff in there, we might need to change the UI a bit to make it more accessible (in fact that whole dictionaries UI needs some improvement, with the confusing priority and numbering stuff which should all just be replaced with drag and drop... which would then allow us to replace the `⋮` with a `⚙` that goes straight to the details popup).

@StefanVukovic99 Anyways, I don't want to block your work just on this, so I will leave it up to you what you'd like to do. Would you like to start by just merging this and then we can improve the dictionary-based settings later?

**Comment by StefanVukovic99** [comment]
I hadn't noticed the dictionary details before. I'd like to make this dictionary-level before merging, should get to that soon. Thank you.

**Comment by StefanVukovic99** [comment]
@toasted-nutbread Have you had a look at this since we changed it to a per-dictionary setting?

**Comment by StefanVukovic99** [comment]
@djahandarie if there is nothing else, this looks mergeable to me.

**Review by djahandarie — APPROVED** [review]

**Pull request merged**
